### PR TITLE
feat: upgrade to Probot v10 / deploy to Vercel

### LINKED ISSUE #329: Discussion: updating the All Contributors App to the latest Probot
Hi 👋🏼 I'm one of the @all-contributors maintainers. I also maintain @probot and @octokit, which the app is build upon. I haven't been very active in this year, but I'd like to help more moving forward.

My first objective is to upgrade the [All Contributors app](https://github.com/marketplace/all-contributors) to the latest version of Probot and to minimize the maintenance overhead.

Here is my preliminary assessment and suggestions on what to do next

### The code

The All Contributors consists of 3 repositories on GitHub

1. https://github.com/all-contributors/all-contributors - source code of website, out of scope for this project
2. https://github.com/all-contributors/all-contributors-cli - source code of the `all-contributors` CLI 
3. https://github.com/all-contributors/all-contributors-bot - source code of the `@allcontributors` app

The `all-contributors` repository is not used by the app. The app is using these exports by the CLI:

- [`generate`](https://github.com/all-contributors/all-contributors-cli/blob/f1dc01d714b75e468465565659ae3e03ddbc9237/src/generate/index.js) - takes the current file content (e.g. from `README.md`), array of contributors and an `options` object. Replaces the content between the `<!-- ALL-CONTRIBUTORS-LIST:` / `-->` tags with the new table of contributors
- [`initBadge`](https://github.com/all-contributors/all-contributors-cli/blob/f1dc01d714b75e468465565659ae3e03ddbc9237/src/init/init-content.js#L23-L25) - update code for all-contributors badge (see [this FAQ question](https://allcontributors.org/docs/en/bot/faq)). 
	I don't understand:
	- why we need the HTML comments around the badge
  - why we need to update the badge at all, why isn't it dynamic in the first place?
- [`initContributorsList`](https://github.com/all-contributors/all-contributors-cli/blob/f1dc01d714b75e468465565659ae3e03ddbc9237/src/init/init-content.js#L41-L60) - I think it initially adds a `## Contributors` section to the README, used when installed initially
- [`addContributorWithDetails`](https://github.com/all-contributors/all-contributors-cli/blob/f1dc01d714b75e468465565659ae3e03ddbc9237/src/contributors/addWithDetails.js) - I think that this is used to update the `.all-contributorsrc` file?

The app code is using `probot` v8 and the `serverless` framework and `aws-sdk`. It's deployed to AWS using `serverless`. It's using `@thundra/core`, but I'm not aware of any public dashboards? I don't think I have access, and it's not mentioned in the [privacy policy](https://github.com/all-contributors/all-contributors-bot/blob/88698016e969ba5a6fc428fac6f45b6104301431/privacy-policy.md). It also sends tracking data to amplitude.com, which is also not mentioned in the privacy policy.

The biggest amount of logic lives in [`src/tasks/processIssueComment`](https://github.com/all-contributors/all-contributors-bot/tree/88698016e969ba5a6fc428fac6f45b6104301431/src/tasks/processIssueComment). The actual probot application function lives [here](https://github.com/all-contributors/all-contributors-bot/blob/88698016e969ba5a6fc428fac6f45b6104301431/src/tasks/processIssueComment/probot-processIssueComment.js#L215-L222). This is the core of the app. Everything else can probably be removed or greatly simplified. [Several checks are done](https://github.com/all-contributors/all-contributors-bot/blob/88698016e969ba5a6fc428fac6f45b6104301431/src/serverless-webhook.js#L62-L119) before invoking the probot app function that should be moved inside it.

My understanding is that the current implementation runs the lambda functions asynchronous to the http request. I don't think that's necessary, because in the most complex path the app creates one comment and a pull request updating two files. Execution time should not exceed 10s.

### The maintenance

As far as I can tell we use the following services for CI, analytics and deployment

- AWS
- thundra.io
- amplitude.com
- Sentry (via Probot)
- CircleCI
- Codecov

For local development and CI, we also use a ton of dependencies

- eslint
- git-cz
- serverless

### The code

I think the `all-contributors` project should consist of 5 repositories:

1. The website
2. The specification: exports a JSON format. Should be super simple to reason about.
3. A library which reads/writes configuration and updates markdown text. Has specification as dependency
4. The CLI. Has library as dependency. Implements the CLI interface
5. The App. Has library as dependency. Implements the GitHub App.

For the app, I think we should remove any trackers and use simple logging as a means to understand how the app is used. I would remove `serverless` and use Probot's new APIs that make it simple to use it in a function. I would suggest we deploy the app to [Vercel](vercel.com/). If we should need async event handling and/or data persistance, I'd suggest we deploy to [Begin](https://begin.com/). Both have a GitHub integration that continuously deploys. Vercel also deploys previews for every pull request. We can register one or several GitHub apps for these preview environments if we want to test the app before deploying to production.

For logging I would use https://logflare.app/ which neatly integrates with Google's Data Studio without any overhead. It makes it trivial to create a public dashboard such as this one: wip.vercel.app/stats

For uptime status monitoring, I would use https://uptimerobot.com. 

### The maintenance

- Replace CircleCI with GitHub Actions
- Remove eslint, just use prettier
- replace yarn with npm. `npm` is the common dominator, yarn folks know how to use `npm`, the other way around this is not necessary the case. 
- In general, remove as many dot files as possible. They all come with a price, mostly to new contributors and new maintainers who don't understand what they are.

There are lots of out-of-date stats in https://github.com/all-contributors/all-contributors/blob/master/MAINTAINERS.md. I would remove everything that cannot be automated. We can revisit the Roadmap/Goals and turn it into actionable items that are reasonable for a small maintenance team.

## Next steps

I want to focus on the app first.

I'm not a fan of rewrites, but in this particular case I think it's the best strategy to move forward. I'll start a pull request where I'll throw away most of the code but will keep the tests for integrity, I'll setup deployment to Vercel and register a new GitHub App for testing while working on it. 

I'll setup the Logflare and Sentry intgrations through [Vercel's integrations](https://vercel.com/integrations). For CI/CD I'll setup GitHub Actions.

### PR DESCRIPTION
**Pull request by gr2m**
fixes #329
fixes #305
fixes #323
closes #327

### TODOs

add tracking via logging

- [x] installations
- [x] errors: known / unknown
- [x] unknown intent
- [x] comment processed
- [x] add contributor: username + contributions array
- [x] When `SENTRY_DSN` is set, do not log the entire error object 👀

### CODE REVIEW

**Comment by gr2m** [comment]
works: https://github.com/gr2m/sandbox/issues/163 / https://github.com/gr2m/sandbox/pull/164

I still need to setup logging and sentry via Vercel, but other than that I'm ready to ship the new app ✨

File: lib/modules/repository.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      if (error.status !== 404) {
+        throw error;
+      }
+
+      throw new ResourceNotFoundError(filePath, this.getFullname());
+    }
+  }
+
+  async getMultipleFiles(filePathsArray) {
+    // TODO: can probably optimise this instead of sending a request per file
```
Comment (Berkmann18) [inline lib/modules/repository.js:67]: Yeah, that would definitely be better, how do you see that happening?

File: lib/process-issue-comment.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  const { who, action, contributions } = parseComment(commentBody);
+
+  if (action !== "add") {
+    commentReply.reply(`I could not determine your intention.`);
+    commentReply.reply(
+      `Basic usage: @all-contributors please add @someone for code, doc and infra`
+    );
+    commentReply.reply(
+      `For other usages see the [documentation](https://allcontributors.org/docs/en/bot/usage)`
+    );
```
Comment (Berkmann18) [inline lib/process-issue-comment.js:20]: Wouldn't merging all three `reply` calls in one (with newlines added of course) be better? I don't see any benefits in having those lines given that the result is only one comment will all this text.

**Review by MarcelRaschke — APPROVED** [review]
review approve

File: lib/parse-comment.js [hunk 3]
```diff
@@ -127,7 +127,7 @@ const plugin = {
 
 nlp.plugin(plugin);
 
-function parseAddComment(message, action) {
+function getMatcher(message) {
   const whoMatched = nlp(message)
     .match(`${action} [.]`)
```
Comment (gr2m) [inline lib/parse-comment.js:132]: `action` is no longer defined, the tests fail now. Was that an unintentional change or do we need to change that line, too @Berkmann18?

Comment (Berkmann18) [reply lib/parse-comment.js:132]: Oops, I missed that bit, it wasn't intentional

**Comment by gr2m** [comment]
all right here goes nothing 🤞🏼

**Comment by gr2m** [comment]
actually let me add a test action workflow before merging :)

**Comment by gr2m** [comment]
happy path works: https://github.com/gr2m/sandbox/issues/163#issuecomment-753381685

I see the logs coming in to the logflare account.

**Pull request merged**